components/ntp_background_images/browser/ntp_sponsored_rich_media_source.cc

```diff
@@ -0,0 +1,168 @@
+                                  scoped_refptr<base::RefCountedMemory>()));
+
+    return;
+  }
+
+  // Allow access.
+  base::ThreadPool::PostTaskAndReplyWithResult(
+      FROM_HERE, {base::MayBlock()},
+      base::BindOnce(&ReadFileToString, *file_path),
+      base::BindOnce(&NTPSponsoredRichMediaSource::ReadFileCallback,
+                     weak_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void NTPSponsoredRichMediaSource::ReadFileCallback(
+    GotDataCallback callback,
+    std::optional<std::string> input) {
+  if (!input) {
+    return std::move(callback).Run(scoped_refptr<base::RefCountedMemory>());
+  }
+
+  std::move(callback).Run(
+      new base::RefCountedBytes(base::as_byte_span(*input)));
+}
+
+std::string NTPSponsoredRichMediaSource::GetMimeType(const GURL& url) {
+  std::string mime_type;
+  const base::FilePath::StringType file_path_extension =
+      base::FilePath::FromUTF8Unsafe(url.path_piece()).Extension();
+  if (!file_path_extension.empty()) {
+    net::GetWellKnownMimeTypeFromExtension(file_path_extension.substr(1),
+                                           &mime_type);
+  }
+
+  return mime_type;
+}
+
+bool NTPSponsoredRichMediaSource::AllowCaching() {
+  return false;
+}
+
+std::string NTPSponsoredRichMediaSource::GetContentSecurityPolicy(
+    network::mojom::CSPDirectiveName directive) {
+  switch (directive) {
+    case network::mojom::CSPDirectiveName::FrameAncestors:
+      return std::string("frame-ancestors ") + kBraveUINewTabURL + ";";
+    case network::mojom::CSPDirectiveName::Sandbox:
+      return "sandbox allow-scripts;";
+    case network::mojom::CSPDirectiveName::DefaultSrc:
+      return "default-src 'none';";
+    case network::mojom::CSPDirectiveName::BaseURI:
+      return "base-uri 'none';";
+    case network::mojom::CSPDirectiveName::FormAction:
+      return "form-action 'none';";
+    case network::mojom::CSPDirectiveName::ScriptSrc:
+      return "script-src 'self';";
+    case network::mojom::CSPDirectiveName::StyleSrc:
+      return "style-src 'self';";
+    case network::mojom::CSPDirectiveName::ImgSrc:
+      return "img-src 'self';";
+    case network::mojom::CSPDirectiveName::MediaSrc:
+      return "media-src 'self';";
+    case network::mojom::CSPDirectiveName::RequireTrustedTypesFor:
+      return "require-trusted-types-for 'script';";
+    case network::mojom::CSPDirectiveName::TrustedTypes:
+      return "trusted-types;";
+    default:
+      // Return empty CSP to avoid inheriting potentially permissive defaults
+      // from content::URLDataSource::GetContentSecurityPolicy().
+      return std::string();
+  }
+}
+
+std::optional<base::FilePath>
+NTPSponsoredRichMediaSource::MaybeGetFilePathForRequestPath(
+    const std::string& request_path) {
+  const NTPSponsoredImagesData* const data =
+      service_->GetBrandedImagesData(false);
+  if (!data) {
+    // No data, deny access.
+    return std::nullopt;
+  }
+
+  const base::FilePath request_path_dir_name =
```

Comment:
do we need to do any extra checks on the request path since we're dealing with untrusted? Could this be used to read other files?

Reply:
We match against the new tab page ads contents folder known as the creative. Files cannot be accessed outside of this folder.

Reply:
See other comment. Tests will be merged tomorrow. Thanks